crypto/multisig/src/lib.rs

```diff
@@ -134,47 +181,63 @@ impl<P: Default + PublicKey + Message> SignerSet<P> {
         }
 
         // Sort and dedup the list of signers and signatures.
         // While the verification code below should be immune to duplicate signers or
         // signatures, the overhead of deduping them is negligible and being
         // extra-safe is a good idea.
-        let mut potential_signers = self.signers.clone();
-        potential_signers.sort();
-        potential_signers.dedup();
+        let mut potential_individual_signers = self.individual_signers.clone();
+        potential_individual_signers.sort();
+        potential_individual_signers.dedup();
+
+        let mut potential_multi_signers = self.multi_signers.clone();
+        potential_multi_signers.sort();
+        potential_multi_signers.dedup();
 
         let mut signatures = multi_sig.signatures.clone();
         signatures.sort_by(|a, b| a.as_ref().cmp(b.as_ref()));
         signatures.dedup();
 
-        // See which signatures which match signers.
-        let mut matched_signers = Vec::new();
-        for signature in signatures.iter() {
-            let matched_signer = potential_signers.iter().find_map(|signer| {
-                signer
-                    .verify(message, signature)
-                    .ok()
-                    .map(|_| signer.clone())
-            });
-            if let Some(matched_signer) = matched_signer {
-                potential_signers.retain(|signer| signer != &matched_signer);
-                matched_signers.push(matched_signer);
+        // Collect individual signer identities that signed the message, and count how
+        // many signers we successfully matched.
+        let mut matched_signer_identities = vec![];
+        let mut num_matches = 0;
+
+        for individual_signer in potential_individual_signers {
+            if signatures
+                .iter()
+                .any(|sig| individual_signer.verify(message, sig).is_ok())
+            {
+                num_matches += 1;
+                matched_signer_identities.push(individual_signer);
+            }
+        }
+
+        // See if any multi-signer sets signed the message.
+        for signer_set in potential_multi_signers {
+            if let Ok(signer_identities) = signer_set.verify(message, multi_sig) {
+                matched_signer_identities.extend(signer_identities);
+                num_matches += 1;
             }
         }
 
         // Did we pass the threshold of verified signatures?
-        if matched_signers.len() < self.threshold as usize {
+        if num_matches < self.threshold as usize {
             return Err(SignatureError::new());
         }
 
-        Ok(matched_signers)
+        // Dedup the list of matched signer identities. We will encounter duplicates in
+        // the case of common signers. (e.g. if a signer set contains a signer
+        // that is also a signer in a nested signer set).
+        matched_signer_identities.sort();
+        matched_signer_identities.dedup();
+
+        Ok(matched_signer_identities)
```

Comment:
I'm a tad confused, shouldn't the check for if the number of valid signers was greater than the threshold occur *after* deduplication?

Reply:
No. The threshold is checked against the number of matches, not the number of identities (in the `if num_matches < self.threshold as usize` line). It's worth noting that there are cases where you can have one match and two identities.
For example, consider a signer set S1 that has threshold=1 and multi_signers=S2. SignerSet S2 has threshold=2 and two public keys.
In this case, the check on S1 would go over the multi identities, and recurse into checking S2 against the public keys provided. If the right two keys were provided, it will return them as the identities and the code [here](https://github.com/mobilecoinfoundation/mobilecoin/blob/3dc49a1513d8aa9ab4d422e7c910d79cecd8da85/crypto/multisig/src/lib.rs#L223-L226) will count it as one match towards the threshold, which is the desired behavior.